lib/hiera/backend/module_data_backend.rb

```diff
@@ -68,13 +67,15 @@ def lookup(key, scope, order_override, resolution_type)
 
         unless module_name
           Hiera.debug("Skipping Module Data backend as this does not look like a module")
-          return no_answer
+          throw_no_such_key
+          return answer
```

Comment:
and here just do as before, `return no_answer` if it's puppet 4 it will throw and if it's 3 it will return nil, just like before

Ditto below, then this PR becomes really small and effects only what the problem is.